feat(core): remove the (^ syntax and make all DOM events bubbling

vsavkin opened the pull request:
BREAKING CHANGE

Before

```
<div (^click)="onEventHandler()">
  <button></button>
</div>
```

After

```
<div (click)="onEventHandler()">
  <button></button>
</div>
```

pkozlowski-opensource commented:
Yes! Out of curiosity - what were the trade-offs here? I mean, why we've decided to not catch bubbling events by default?

mhevery commented:
@pkozlowski-opensource The change is that event decides if it is bubbling or not. In practice it turns out that you want all user events (mouse, keys, etc) to bubble, but non of the app events (open,close,onChange, etc) to bubble.

So we are going the route of event declaring if it wants to bubble or not.

mhevery commented:
Related: #2296

vicb commented on modules/angular2/src/core/render/dom/events/hammer_gestures.dart:
```diff
@@ -18,10 +18,7 @@ class HammerGesturesPlugin extends HammerGesturesPluginCommon {
-      bool shouldSupportBubble) {
-    if (shouldSupportBubble) throw new BaseException(
-        'Hammer.js plugin does not support bubbling gestures.');
+  addEventListener(Element element, String eventName, Function handler) {
```
Hammer events do not bubble by default - because they are not using DOM events.

It is possible to configure Hammer events to bubble (by using DOM event, it comes with a perf penalty) ? This could peobably be done in a follow up PR.

gkalpak commented on modules/angular2/src/core/render/dom/compiler/property_binding_parser.ts:
```diff
@@ -12,14 +12,13 @@ import {dashCaseToCamelCase} from '../util';
+// Group 8 = identifier inside ()
 var BIND_NAME_REGEXP =
-    /^(?:(?:(?:(bind-)|(var-|#)|(on-)|(onbubble-)|(bindon-))(.+))|\[\(([^\)]+)\)\]|\[([^\]]+)\]|\(([^\)]+)\))$/g;
+    /^(?:(?:(?:(bind-)|(var-|#)|(on-)|(bindon-))(.+))|\[\(([^\)]+)\)\]|\[([^\]]+)\]|\(([^\)]+)\))$/g;
```
Shouldn't the `^` parts also be removed from the RegExp ?

PascalPrecht commented:
Will custom events bubble up by default too then?

vsavkin commented:
@PascalPrecht custom dom events can bubble, but synthetic events won't.

mary-poppins commented:
Merging PR #3864 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-3864](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-3864).

vsavkin committed:
feat(core): remove the (^ syntax and make all DOM events bubbling

BREAKING CHANGE

Before
<div (^click)="onEventHandler()">
  <button></button>
</div>

After
<div (click)="onEventHandler()">
  <button></button>
</div>

mary-poppins commented:
Merging PR #3864 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-3864](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-3864).
